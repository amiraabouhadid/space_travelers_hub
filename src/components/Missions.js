import React from 'react';
import { useSelector, shallowEqual } from 'react-redux';
import Mission from './Mission';

const Missions = () => {
  const missions = useSelector((state) => state.missions, shallowEqual);

  return (
    <div className="mx-8 my-8 px-5 py-5 text-left">
      <table className="table table-striped table-bordered table-responsive">
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th> join mission</th>
          </tr>
        </thead>
        <tbody>

          {missions.map((mission) => (
            <tr key={mission.id}>
              <Mission
                description={mission.description}
                name={mission.name}
                id={mission.id}
              />
            </tr>
          ))}

        </tbody>
      </table>
    </div>
  );
};
export default Missions;
