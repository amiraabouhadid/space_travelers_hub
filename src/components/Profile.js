import React from 'react';
import PropTypes from 'prop-types';
import RocketButton from './RocketButton';

const Profile = (props) => {
  const { missions, rockets } = props;

  const displayer = (container, message) => {
    const displayble = container.filter((content) => content.reserved);
    let displayed;

    if (displayble.length > 0) {
      displayed = displayble.map((item) => (
        <tr key={item.id}>
          <td className="p-3 flex justify-between items-center">
            {item.name}
            {message.includes('rockets') && <RocketButton rocket={item} />}
          </td>
        </tr>
      ));
    } else {
      displayed = (
        <tr>
          <td className="p-3 italic">
            {message}
          </td>
        </tr>
      );
    }

    return displayed;
  };

  return (
    <div className="p-5 row">
      <div className="col-6">
        <h2>
          My Missions
        </h2>
        <table className="table table-hover table-bordered">
          <tbody>
            {displayer(missions, 'No missions joined')}
          </tbody>
        </table>
      </div>
      <div className="col-6">
        <h2>
          My Rockets
        </h2>
        <table className="table table-hover table-bordered">
          <tbody>
            {displayer(rockets, 'No rockets reserved')}
          </tbody>
        </table>
      </div>
    </div>
  );
};

Profile.propTypes = {
  missions: PropTypes.instanceOf(Array).isRequired,
  rockets: PropTypes.instanceOf(Array).isRequired,
};

export default Profile;
