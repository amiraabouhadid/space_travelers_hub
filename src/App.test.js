import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import Store from './redux/configureStore';
import Missions from './components/Missions';
import Profile from './components/Profile';

describe('Profile page component', () => {
  test('renders profile headers', () => {
    const missions = [{ id: '123', name: 'Falcon', description: 'a dummy text created for testing purposes' }];
    const rockets = [{
      id: '1329', name: 'myStar', description: 'second dummy text', image: 'cool pic',
    }];
    render(
      <Provider store={Store}>
        <Profile missions={missions} rockets={rockets} />
      </Provider>,
    );
    const profilePageHeaders = screen.getAllByText(/\b(my missions)\b|\b(my rockets)\b/i);
    expect(profilePageHeaders).toHaveLength(2);
  });
});

describe('Missions page component', () => {
  test('renders  missions table header ', () => {
    const missions = [{ id: '123', name: 'Falcon', description: 'a dummy text created for testing purposes' }];

    render(
      <Provider store={Store}>
        <Missions missions={missions} />
      </Provider>,
    );
    const missionsTableHeader = screen.getAllByText(/\b(^mission)\b|\b(description)\b|\b(status)\b/i);
    expect(missionsTableHeader).toHaveLength(3);
  });
  test('renders  mission  name', () => {
    const missions = [{ id: '123', name: 'Falcon', description: 'a dummy text created for testing purposes' }];

    render(
      <Provider store={Store}>
        <Missions missions={missions} />
      </Provider>,
    );
    const missionName = screen.getAllByText(/\bfalcon\b/i);
    expect(missionName).toHaveLength(1);
  });
  test('renders  mission  description ', () => {
    const missions = [{ id: '123', name: 'Falcon', description: 'a dummy text created for testing purposes' }];

    render(
      <Provider store={Store}>
        <Missions missions={missions} />
      </Provider>,
    );
    const missionDescription = screen.getAllByText(/\ba dummy text created for testing purposes\b/i);
    expect(missionDescription).toHaveLength(1);
  });
  test('renders  mission  status badge before joining mission', () => {
    const missions = [{ id: '123', name: 'Falcon', description: 'a dummy text created for testing purposes' }];

    render(
      <Provider store={Store}>
        <Missions missions={missions} />
      </Provider>,
    );
    const missionDescription = screen.getAllByText(/\bnot a member\b/i);
    expect(missionDescription).toHaveLength(1);
  });
  test('renders mission join mission button before joining mission', () => {
    const missions = [{ id: '123', name: 'Falcon', description: 'a dummy text created for testing purposes' }];

    render(
      <Provider store={Store}>
        <Missions missions={missions} />
      </Provider>,
    );
    const missionDescription = screen.getAllByText(/\bjoin mission\b/i);
    expect(missionDescription).toHaveLength(1);
  });
});
