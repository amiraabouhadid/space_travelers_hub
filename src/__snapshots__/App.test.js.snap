// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions page component changes status badge and join button when mission is joined  1`] = `
<div
  className="mx-8 my-8 px-5 py-5 text-left"
>
  <table
    className="table table-striped table-bordered table-responsive"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
          Join
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="font-bold"
        >
          Falcon
        </td>
        <td>
          a dummy text created for testing purposes
        </td>
        <td
          className="align-middle"
        >
          <span
            className="badge badge-secondary bg-secondary"
          >
            NOT A MEMBER
          </span>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="btn btn-light border"
            onClick={[Function]}
            type="submit"
          >
            Join mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Missions page component changes status badge when mission is joined  1`] = `
<div
  className="mx-8 my-8 px-5 py-5 text-left"
>
  <table
    className="table table-striped table-bordered table-responsive"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
          Join
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="font-bold"
        >
          Falcon
        </td>
        <td>
          a dummy text created for testing purposes
        </td>
        <td
          className="align-middle"
        >
          <span
            className="badge badge-secondary bg-secondary"
          >
            NOT A MEMBER
          </span>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="btn btn-light border"
            onClick={[Function]}
            type="submit"
          >
            Join mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Missions page component renders correctly 1`] = `
<div
  className="mx-8 my-8 px-5 py-5 text-left"
>
  <table
    className="table table-striped table-bordered table-responsive"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
          Join
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="font-bold"
        >
          Falcon
        </td>
        <td>
          a dummy text created for testing purposes
        </td>
        <td
          className="align-middle"
        >
          <span
            className="badge badge-secondary bg-secondary"
          >
            NOT A MEMBER
          </span>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="btn btn-light border"
            onClick={[Function]}
            type="submit"
          >
            Join mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Profile page component adds joined missions to profile when join mission button is clicked  1`] = `
<div
  className="p-5 row"
>
  <div
    className="col-6"
  >
    <h2>
      My Missions
    </h2>
    <table
      className="table table-hover table-bordered"
    >
      <tbody />
    </table>
  </div>
  <div
    className="col-6"
  >
    <h2>
      My Rockets
    </h2>
    <table
      className="table table-hover table-bordered"
    >
      <tbody />
    </table>
  </div>
</div>
`;

exports[`Profile page component renders correctly 1`] = `
<div
  className="p-5 row"
>
  <div
    className="col-6"
  >
    <h2>
      My Missions
    </h2>
    <table
      className="table table-hover table-bordered"
    >
      <tbody />
    </table>
  </div>
  <div
    className="col-6"
  >
    <h2>
      My Rockets
    </h2>
    <table
      className="table table-hover table-bordered"
    >
      <tbody />
    </table>
  </div>
</div>
`;

exports[`Rockets page component add reserved badge and cancel reservation button when the rocket is booked 1`] = `
<main
  className="w-11/12 flex flex-col items-center p-8 gap-4 rockets-section"
>
  <div
    className="flex items-center w-full gap-4 rocket-container"
  >
    <img
      alt="Falcon"
      className="rocket-image"
    />
    <div>
      <h2
        className="text-xl font-semibold"
      >
        Falcon
      </h2>
      <p
        className="text-md"
      >
        a dummy text created for testing purposes
      </p>
      <button
        className="text-sm border-1 border-blue-500 text-white bg-blue-500 rounded-md p-2"
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</main>
`;

exports[`Rockets page component renders correctly 1`] = `
<main
  className="w-11/12 flex flex-col items-center p-8 gap-4 rockets-section"
>
  <div
    className="flex items-center w-full gap-4 rocket-container"
  >
    <img
      alt="Falcon Heavy"
      className="rocket-image"
    />
    <div>
      <h2
        className="text-xl font-semibold"
      >
        Falcon Heavy
      </h2>
      <p
        className="text-md"
      >
        a dummy text created for testing purposes
      </p>
      <button
        className="text-sm border-1 border-blue-500 text-white bg-blue-500 rounded-md p-2"
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</main>
`;
